src/core-manager/core-replication-state.js

```diff
@@ -0,0 +1,395 @@
+    return 0
+  }
+  /**
+   * Returns whether this peer wants block at `index`. Defaults to `true` for
+   * all blocks
+   * @param {number} index
+   */
+  want(index) {
+    return this.#wants ? this.#wants.get(index) : true
+  }
+  /**
+   * @param {number} index
+   */
+  wantWord(index) {
+    if (this.#wants) return getBitfieldWord(this.#wants, index)
+    return 2 ** 32 - 1
+  }
+}
+
+/**
+ * Derive count for each peer: "want"; "have"; "wanted". There is definitely a
+ * more performant and clever way of doing this, but at least with this
+ * implementation I can understand what I am doing.
+ *
+ * @param {InternalState} coreState
+ *
+ * @private
+ * Only exporteed for testing
+ */
+export function deriveState(coreState) {
+  const peerIds = ['local', ...coreState.remoteStates.keys()]
+  const peers = [coreState.localState, ...coreState.remoteStates.values()]
+
+  /** @type {PeerSimpleState[]} */
+  const peerStates = new Array(peers.length)
+  const length = coreState.length || 0
+  for (let i = 0; i < peerStates.length; i++) {
+    peerStates[i] = { want: 0, have: 0, wanted: 0, missing: 0 }
+  }
+  const haves = new Array(peerStates.length)
+  let want = 0
+  for (let i = 0; i < length; i += 32) {
+    const truncate = 2 ** Math.min(32, length - i) - 1
+    let someoneHasIt = 0
+    for (let j = 0; j < peers.length; j++) {
+      haves[j] = peers[j].haveWord(i) & truncate
+      someoneHasIt |= haves[j]
+      peerStates[j].have += bitCount(haves[j])
+    }
+    let someoneWantsIt = 0
+    for (let j = 0; j < peers.length; j++) {
+      // A block is a want if:
+      //   1. The peer wants it
+      //   2. They don't have it
+      //   3. Someone does have it
+      const wouldLikeIt = peers[j].wantWord(i) & ~haves[j]
+      want = wouldLikeIt & someoneHasIt
+      someoneWantsIt |= want
+      peerStates[j].want += bitCount(want)
+      peerStates[j].missing += bitCount(wouldLikeIt & ~someoneHasIt & truncate)
+    }
+    for (let j = 0; j < peerStates.length; j++) {
+      // A block is wanted if:
+      //   1. Someone wants it
+      //   2. The peer has it
+      const wanted = someoneWantsIt & haves[j]
+      peerStates[j].wanted += bitCount(wanted)
+    }
+  }
+  /** @type {DerivedState} */
+  const derivedState = {
+    coreLength: length,
+    localState: peerStates[0],
+    remoteStates: {},
+  }
+  for (let j = 1; j < peerStates.length; j++) {
+    const peerState = /** @type {RemotePeerSimpleState} */ (peerStates[j])
+    peerState.connected = peers[j].connected
+    derivedState.remoteStates[peerIds[j]] = peerState
+  }
+  return derivedState
+}
+
+/** @param {number} n */
+function bitCount(n) {
+  n = n - ((n >> 1) & 0x55555555)
+  n = (n & 0x33333333) + ((n >> 2) & 0x33333333)
+  return (((n + (n >> 4)) & 0xf0f0f0f) * 0x1010101) >> 24
+}
+
+/**
+ *
+ * @param {RemoteBitfield} bitfield
+ * @param {number} index
+ */
+function getBitfieldWord(bitfield, index) {
```

Comment:
mind briefly elaborating on what "word" means in this context?

Reply:
It means "32-bit integer", since the bitfield is a `UInt32Array` - this reads a single entry from that array. I'll add a code comment.